src/slptokens/token.h

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2019 Greg Griffith
+// Copyright (c) 2019 The Bitcoin Unlimited developer
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_SLP_TOKEN_H
+#define BITCOIN_SLP_TOKEN_H
+
+#include "script/script.h"
+#include "serialize.h"
+
+#include <stdint.h>
+#include <vector>
+
+enum SLP_TX_TYPE : uint8_t
+{
+    SLP_NULL = 0,
+    SLP_GENESIS,
+    SLP_MINT,
+    SLP_SEND,
+    SLP_COMMIT
+};
+
+class CSLPToken
+{
+    // using vectors for the variable length fields is not ideal, but it is the easiest way to do it
+private:
+    // SLP Fields
+    uint16_t token_type; // 1-2 bytes
+    SLP_TX_TYPE tx_type; // 4-6 bytes
+
+    // NOT IN GENESIS
+    std::vector<uint8_t> token_id; // 32 bytes, token genesis transaction hash
+
+    // NOT IN SEND
+    uint8_t mint_baton_vout; //(0 bytes or 1 byte between 0x02-0xff), if 0 there is no mint baton
+
+    // ONLY IN SEND
+    std::vector<uint64_t> token_output_quantitys; // this can be a max size of 19
+
+    // ONLY IN GENESIS
+    std::vector<uint8_t> token_ticker; // (0 or more bytes, suggested utf-8)
+    std::vector<uint8_t> token_name; // (0 or more bytes, suggested utf-8)
+    std::vector<uint8_t> token_document_url; // (0 or more bytes, suggested ascii)
+    std::vector<uint8_t> token_document_hash; // 32 bytes
+    uint8_t decimals;
+    uint64_t initial_token_mint_quantity;
+    // ONLY IN SEND
+    uint64_t additional_token_quantity;
+
+public:
+    // UTXO Fields
+    int nHeight;
+
+private:
+    bool DetermineDynamicSize(const CScript &scriptPubKeyIn, size_t &index, uint32_t &return_size);
```

Comment:
even though these are private, doxygen-comment them anyway so that the reviewer knows what they are supposed to do.